aries_cloudagent/transport/pack_format.py

```diff
@@ -218,3 +300,106 @@ def get_recipient_keys(self, message_body: Union[str, bytes]) -> List[str]:
+
+        LOGGER.debug("HIT V2 PACK FORMAT .parse_message()")
+        LOGGER.debug(message_body)
+
+        receipt = MessageReceipt()
+        receipt.in_time = time_now()
+        receipt.raw_message = message_body
+        receipt.didcomm_version = DIDCommVersion.v2
+
+        message_dict = None
+        message_json = message_body
+
+        if not message_json:
+            raise WireFormatParseError("Message body is empty")
+
+        try:
+            message_dict = json.loads(message_json)
+        except ValueError:
+            raise WireFormatParseError("Message JSON parsing failed")
+        if not isinstance(message_dict, dict):
+            raise WireFormatParseError("Message JSON result is not an object")
+
+        wallet = session.inject_or(BaseWallet)
+        if not wallet:
+            LOGGER.error("No wallet found")
+            raise WalletNotFoundError()
+
+        # packed messages are detected by the absence of type
+        if "type" not in message_dict:
+            try:
+                message_unpack = await messaging.unpack(message_json)
+            except CryptoServiceError:
+                LOGGER.debug("Message unpack failed, falling back to JSON")
+            else:
+                # Set message_dict to be the dictionary that we unpacked
+                message_dict = message_unpack.message
+
+        if message_unpack.sender_kid:
+            receipt.sender_verkey = message_unpack.sender_kid.split("#")[0]
+            receipt.recipient_verkey = message_unpack.recipient_kid.split("#")[0]
+            pass
+
+        thid = message_dict.get("thid")
+        receipt.thread_id = thid or message_dict.get("id")
+
+        receipt.parent_thread_id = message_dict.get("pthid")
+
+        # handle transport decorator
+        receipt.direct_response_mode = message_dict.get("return_route")
+
+        LOGGER.debug("Expanded message: %s", message_dict)
+
+        return message_dict, receipt
+
+    async def encode_message(
+        self,
+        session: ProfileSession,
+        message_json: Union[str, bytes],
+        recipient_keys: Sequence[str],
+        routing_keys: Sequence[str],
+        sender_key: str,
+    ) -> Union[str, bytes]:
+        """Encode an outgoing message for transport.
+
+        Args:
+            session: The profile session for providing wallet access
+            message_json: The message body to serialize
+            recipient_keys: A sequence of recipient verkeys
+            routing_keys: A sequence of routing verkeys
+            sender_key: The verification key of the sending agent
+
+        Returns:
+            The encoded message
+
+        Raises:
+            MessageEncodeError: If the message could not be encoded
+
+        """
+        messaging = session.inject(DIDCommMessaging)
+
+        if sender_key and recipient_keys:
+            message = await messaging.pack(
+                message=message_json,
+                to=recipient_keys[0],
```

Comment:
Should the `pack` method of DMP that we're using here support multiple recipients?

Reply:
It probably should. Another problem that I ran into is that it implicitly forward wraps the message, even if we are responding on the same connection. Something I planned to look at later